### PR TITLE
CI: Add a new run to check whether `datafusion-cli` lock file is up-to-date

### LINKED ISSUE #3744: Prevent Cargo.lock for datafusion-cli being out-of-date
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Since the only `Cargo.lock` that is checked in is the `datafusion-cli`'s one, we seem to be forgetting about updating [it sometimes](https://github.com/apache/arrow-datafusion/pulls?q=is%3Apr+Cargo.lock+is%3Aclosed).

**Describe the solution you'd like**
Since we already have infrastructure (a lint job), I think it should be relatively simple to add a step that checks whether the `Cargo.lock` is up-to-date or not. It could then fail on PRs that change `Cargo.toml` (add new stuff to other crates), and let the PR authors know that they need to update the lock as well.

**Describe alternatives you've considered**
Leaving as is.

**Additional context**

### PR DESCRIPTION
**Pull request by isidentical**
# Which issue does this PR close?

Closes #3744.

 # Rationale for this change

This adds a new script (that only runs on the CI, so a classic `./dev/rust_lint.sh` doesn't waste its time on this check) that ensures the lock file for the datafusion-cli is up-to-date.

### CODE REVIEW

**Comment by isidentical** [comment]
Want to make sure first that it fails (since the lock is out-of-date now) and I will also include the regenerated version, though the CI does not start without a maintainer's approval.

File: .github/workflows/rust.yml
Comment (alamb) [inline .github/workflows/rust.yml:64]: 👍 

I ran this locally and it worked great for me, FWIW

File: .gitignore
Comment (alamb) [inline .gitignore:88]: 🧹

**Comment by alamb** [comment]
@isidentical  it appears this test did not fail so I am not sure what the status of the PR is -- marking it as "draft" (to help keep the list of PRs ready to review small). Please mark it as ready to review when it is.

Thanks!

**Comment by isidentical** [comment]
@alamb I already have confirmed that it worked with the action failing (and then submitted the changes on the Cargo.lock as well, so that they won't fail now). The PR should be good for a review as is (if I am not missing anything).

**Comment by alamb** [comment]
Cool -- thanks @isidentical  -- I couldn't tell from the github PR history.

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Thanks again @isidentical

**Pull request merged**
